pkg/streamingpromql/optimize/ast/reduce_matchers.go

```diff
@@ -1,62 +1,100 @@
+	return &ReduceMatchers{
+		attempts: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_mimir_query_engine_reduce_matchers_attempted_total",
+			Help: "Total number of queries that the optimization pass has attempted to reduce matchers for.",
+		}),
+		success: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_mimir_query_engine_reduce_matchers_success_total",
+			Help: "Total number of queries where the optimization pass has been able to reduce matchers for.",
+		}),
+		logger: logger,
+	}
 }
 
-func (p concreteLookupPlan) ScanMatchers() []*labels.Matcher {
-	return p.scanMatchers
+// ReduceMatchers deduplicates matchers from vector or matrix selectors, removes matchers that
+// select for all non-empty values if a more selective matcher for the same label name already
+// exists, and removes matchers that select a superset of other matchers. Input order of matchers
+// is NOT preserved in the rewritten expression.
+type ReduceMatchers struct {
+	attempts prometheus.Counter
+	success  prometheus.Counter
+	logger   log.Logger
 }
 
-func (p concreteLookupPlan) IndexMatchers() []*labels.Matcher {
-	return p.indexMatchers
+func (c *ReduceMatchers) Name() string {
+	return "Reduce matchers"
 }
 
-// PlanIndexLookup takes an index.LookupPlan and removes matchers that match only supersets of other matchers.
-// It does not modify plans if the input plan has any scan matchers. It does not guarantee matcher order of the output plan.
-func (p MatcherReducerPlanner) PlanIndexLookup(ctx context.Context, inPlan index.LookupPlan, _, _ int64) (index.LookupPlan, error) {
-	// For simplicity, we don't process plans with scan matchers
-	if planningDisabled(ctx) || len(inPlan.ScanMatchers()) > 0 {
-		return inPlan, nil
-	}
-	// If there's only one matcher, don't try to do any optimizations
-	if len(inPlan.IndexMatchers()) <= 1 {
-		return inPlan, nil
+func (c *ReduceMatchers) Apply(ctx context.Context, root parser.Expr) (parser.Expr, error) {
+	spanlog := spanlogger.FromContext(ctx, c.logger)
+	c.attempts.Inc()
+
+	matchersReduced := false
+	parser.Inspect(root, func(node parser.Node, _ []parser.Node) error {
+		switch expr := node.(type) {
+		case *parser.VectorSelector:
+			retained, dropped := reduceMatchers(expr.LabelMatchers)
+
+			if len(dropped) > 0 {
+				expr.LabelMatchers = retained
+				matchersReduced = true
+				spanlog.DebugLog(
+					"msg", "dropped matchers for vector selector",
+					"retained", util.MatchersStringer(retained),
+					"dropped", util.MatchersStringer(dropped),
+				)
+			}
+		case *parser.MatrixSelector:
+			retained, dropped := reduceMatchers(expr.VectorSelector.(*parser.VectorSelector).LabelMatchers)
+
+			if len(dropped) > 0 {
+				expr.VectorSelector.(*parser.VectorSelector).LabelMatchers = retained
+				matchersReduced = true
+				spanlog.DebugLog(
+					"msg", "dropped matchers for vector selector",
```

Comment:
[nit] Should this be "matrix selector"? Or perhaps drop the "vector" from both this message and the one above?